cabal-install/Distribution/Client/TargetProblem.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Distribution.Client.TargetProblem
+  ( TargetProblem,
+    ExtensibleTargetProblem (..),
+    reportTargetProblems,
+    renderTargetProblem,
+    commonTargetProblem,
+    noneEnabledTargetProblem,
+    noTargetsProblem,
+    customTargetProblem,
+  )
+where
+
+import Distribution.Client.CmdErrorMessages
+  ( renderTargetProblemCommon,
+    renderTargetProblemNoTargets,
+    renderTargetProblemNoneEnabled,
+  )
+import Distribution.Client.Compat.Prelude
+import Distribution.Client.ProjectOrchestration
+  ( TargetProblemCommon (..),
+  )
+import Distribution.Client.ProjectPlanning
+  ( AvailableTarget,
+  )
+import Distribution.Client.TargetSelector
+  ( TargetSelector,
+  )
+import Distribution.Simple.Utils
+  ( die',
+  )
+import Distribution.Verbosity
+  ( Verbosity,
+  )
+import Prelude ()
+
+-- | Type alias for a 'TargetProblem' with no user-defined problems/errors.
+--
+-- Can use the utilities below for reporting/rendering problems.
+type TargetProblem = ExtensibleTargetProblem ()
+
+-- | The various error conditions that can occur when matching a
+-- 'TargetSelector' against 'AvailableTarget's which can be extended
+-- with command specific target problems as described by 'e'.
+data ExtensibleTargetProblem e
```

Comment:
Can `TargetProblemCommon` and `ExtensibleTargetProblem` be merged into own?

Reply:
That would be doable, would just require some more code shuffling. One question would be if we want to inline all of the cases of `TargetProblemCommon` into the top level `ExtensibleTargetProblem` or if we want to keep the hierarchy as is.

At the least we could move the `TargetProblemCommon` definition into `TargetProblem.hs`, I'd prefer to leave the change of flattening the hierarchy (if that's what we decide to do) to a follow-up PR to keep this one as simply extracting the repeated code into one location.

Reply:
Please shuffle code.

- `TargetProblemCommon` is a good name
- `ExtensibleTargetProblem` is bad name
    - `ExtensibleTargetProblemNoTarget` is also bad. First we rename it from `TargetProblemNoTarget` to then rename back to `TargetProblemNoTarget` again.

---

This commit is good start. But I won't merge it. Please at least try `TargetProblemCommon` extraction as well.

Reply:
See most recent commit, to summarize:
- Rename `ExtensibleTargetProblem` -> `TargetProblem` (and change alias)
- Shorten data constructor names in `TargetProblem`
- Extracts `TargetProblemCommon` into `TargetProblem.hs`

PTAL!